### PR TITLE
several changes on vochain and testing

### PR DESCRIPTION
**Pull request by p4u**
The main change (the point to start this PR) was to replace the current Protobuf message "Vote" by an internal GoLang type.

There is no need to declare a protobuf model on this type and it is messy with other protobuf types such as Envelope.

The only reason to have this type as Protobuf model, was to be able to serialize it in order to compute the Hash, which is stored into the blockchain state. However this has been replaced by a new method Hash() that serializes the most rellevant fields of the Vote type (which are already byte arrays).

On this point, the state.AddVote() method has been changed, instead of just storing the vote hash, we also store now the Weight and VotePackage. Thus it will make possible to compute an election results using only state information (without the Indexer).

Then I got into the rabit hole... fixing and improving testing and other stuff.

Signed-off-by: p4u <pau@dabax.net>

### CODE REVIEW

File: api/accounts.go [hunk 1]
```diff
@@ -211,7 +211,7 @@ func (a *API) accountSetHandler(msg *apirest.APIdata, ctx *httprouter.HTTPContex
 		} else {
 			resp.MetadataURL = a.storage.URIprefix() + cid
 		}
-		if cid != metadataCID {
+		if strings.TrimPrefix(cid, "ipfs://") != strings.TrimPrefix(metadataCID, "ipfss://") {
```
Comment (jordipainan) [inline api/accounts.go]: Is `ipfss://` correct here? Seems a typo

File: test/api_test.go [hunk 2]
```diff
@@ -165,23 +165,24 @@ func TestAPIcensusAndVote(t *testing.T) {
...
-	   qt.Assert(t, code, qt.Equals, 200)
-	   v2 := &api.Vote{}
-	   err = json.Unmarshal(resp, v2)
-	   qt.Assert(t, err, qt.IsNil)
-	   qt.Assert(t, v2.VoteID.String(), qt.Equals, v.VoteID.String())
-	   qt.Assert(t, v2.BlockHeight, qt.Equals, uint32(3))
-	   qt.Assert(t, v2.TransactionIndex, qt.Equals, uint32(0))
-	   qt.Assert(t, v2.VoterID.String(), qt.Equals, voterKey.AddressString())
-	*/
+	//  This test is disabled because the indexer is not properly set up in the test environment.
```
Comment (jordipainan) [inline test/api_test.go]: Can you please add the `TODO` keyword so we can later filter all TODO's ?

File: test/api_test.go [hunk 3]
```diff
@@ -202,6 +203,7 @@ func TestAPIaccount(t *testing.T) {
 	waitUntilHeight(t, c, 1)
 
 	// create a new account
+
```
Comment (jordipainan) [inline test/api_test.go]: Unnecessary empty line

File: vochain/app.go [hunk 4]
```diff
@@ -554,8 +579,8 @@ func (app *BaseApplication) Query(req abcitypes.RequestQuery) abcitypes.Response
 	return abcitypes.ResponseQuery{}
 }
 
-// EndBlock updates the app height and timestamp at the end of the current block
-func (app *BaseApplication) EndBlock(req abcitypes.RequestEndBlock) abcitypes.ResponseEndBlock {
+// fnEndBlockDefault updates the app height and timestamp at the end of the current block
+func (app *BaseApplication) fnEndBlockDefault(req abcitypes.RequestEndBlock) abcitypes.ResponseEndBlock {
```
Comment (jordipainan) [inline vochain/app.go:583]: Why the `default` naming ?

Comment (p4u) [reply vochain/app.go:583]: Do you have a better suggestion?

Comment (jordipainan) [reply vochain/app.go:583]: I would just get rid of `default` and use `fnEndBlock` which is the default, and then you have the option of setting it for changing the behaviour

Comment (p4u) [reply vochain/app.go:583]: This collides with the definition on BaseApplication

Comment (jordipainan) [reply vochain/app.go:583]: ok, so ignore my comment. I do not have a better suggestion.

File: test/api_test.go (hunk 2 above)
Comment (p4u) [reply test/api_test.go]: Actually, I re-enabled this test, the comment was just a garbage.

**Review by jordipainan — APPROVED** [review]

**Pull request merged**